test/metabase/lib/datascript_test.cljc

```diff
@@ -0,0 +1,1100 @@
+                                           {:metadata.column/source orders-join
+                                            :metadata.column/mirror
+                                            {:metadata.field/id (meta/id :orders :user-id)}}}
+                                          {:mbql/series 1
+                                           :mbql/ref
+                                           {:metadata.column/source people-join2
+                                            :metadata.column/mirror
+                                            {:metadata.field/id (meta/id :people :id)}}}]}]}
+                people-join2))))))
+
+(deftest ^:parallel multi-stage-test
+  (testing "multi-stage query"
+    (let [query       (-> (ld/query meta/metadata-provider (meta/table-metadata :orders))
+                          (ld/aggregate -1 (ld/agg-count))
+                          (ld/breakout -1 (ld/with-temporal-bucketing
+                                            [:metadata.field/id (meta/id :orders :created-at)]
+                                            :month))
+                          ld/append-stage)
+          stage0      (:mbql.query/stage0 query)
+          stage0-cols [{:metadata.column/source stage0
+                        :metadata.column/mirror (-> stage0 :mbql.stage/breakout first)}
+                       {:metadata.column/source stage0
+                        :metadata.column/mirror (-> stage0 :mbql.stage/aggregation first)}]]
+      (testing "stage 0 returned-columns have the stage as their source"
+        (is (=? stage0-cols
+                (ld/returned-columns query 0))))
+      (testing "stage 1 visible-columns has those columns"
+        (is (=? [{:lib/type                             :metadata/column-group
+                  :metabase.lib.column-group/group-type :group-type/main
+                  :metabase.lib.column-group/columns    stage0-cols}]
+                (ld/visible-column-groups query -1))))
+
+      (testing "pieces in latter stage"
+        (let [[created-at
+               count-col] (->> query :mbql.query/stage0 :mbql.stage/returned (sort-by :mbql/series))
+              filtered    (ld/filter query -1 (ld/expr :> count-col 100))]
+          (is (=? {:mbql.query/stage
+                   {:mbql.stage/filter
+                    [{:mbql.clause/operator :>
+                      :mbql.clause/argument [{:mbql/series 0
+                                              :mbql/ref    (->> filtered
+                                                                :mbql.query/stage0
+                                                                :mbql.stage/returned
+                                                                (sort-by :mbql/series)
+                                                                last)}
+                                             {:mbql/series 1, :mbql/lit 100}]
+                      :mbql/series          0}]}}
+                  filtered)))))))
+
+(defn- transitive-closure
```

Comment:
Some helpers for testing that `remove-clause` leaves the database clean - no dangling refs, no orphaned things detached from the query that didn't get deleted.